### PR TITLE
Upgrade to React 18

### PR DESCRIPTION
**Pull request by sapkra**
### What does it do?

* updates `react` to 18
* updates `react-dom` to 18
* updates `react-is` to 18
* updates `react-test-renderer` to 18
* updates `react-query` to 3.39.2 (so it officially supports react18)
* removes `@testing-library/react-hooks`
* updates `@testing-library/react` to 13.4
* Converts hook tests to use RTL `renderHook`

### Why is it needed?

- use up-to-date version of important library
- improve rendering performance
- add compatibility for React 18 monorepos
- Means we can update a lot more libraries.

### Related issue(s)/PR(s)

https://feedback.strapi.io/feature-requests/p/add-support-to-react-18

### CODE REVIEW

**Comment by sapkra** [comment]
I fixed almost all failing tests, but there a still a few I need help with.

**Comment by joshuaellis** [comment]
Won't all the hook tests need updating because `@testing-library/react-hooks` does not support `r18`? 🤔

File: docs/package.json [hunk 1]
```diff
@@ -19,8 +19,8 @@
     "@mdx-js/react": "^1.6.22",
     "clsx": "^1.1.1",
     "prism-react-renderer": "^1.3.3",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2"
+    "react": "^18.2.0",
```
Comment (joshuaellis) [inline docs/package.json:22]: Can you revert this change? Our CI is skipping because you've edited the docs (not ideal)

Comment (sapkra) [reply docs/package.json]: I removed it from the commit.

Comment (sapkra) [reply docs/package.json]: But I introduced another problem - I'm on it.

**Comment by sapkra** [comment]
> Won't all the hook tests need updating because `@testing-library/react-hooks` does not support `r18`? 🤔

I refactored it, it was more than I thought. It would be great to get some feedback on this. 🙂

**Review by remidej — COMMENTED** [review]
As much as I would want React 18 in Strapi, I don't think it would be possible without causing problems with the community plugins. Many plugins use libraries that expect React ^17, running these with React 18 will create issues. I'm afraid we'll need to wait for Strapi v5 as this looks like a breaking change

**Comment by joshuaellis** [comment]
> As much as I would want React 18 in Strapi, I don't think it would be possible without causing problems with the community plugins. Many plugins use libraries that expect React ^17, running these with React 18 will create issues. I'm afraid we'll need to wait for Strapi v5 as this looks like a breaking change

I'd like to discuss more about this – react 18 is backwards compatible with react 17 assuming you're not using strict mode, which we're not? So why do you think updating to react 18 would cause these breaking changes? 🤔  If anything it would only cause potential install errors with npm without `--legacy-peer-deps` which i imagine is already a problem before this looking at my terminal at least.

**Comment by sapkra** [comment]
We are using a react 18 resolution for several months now and we never encountered an compatibility issue.

**Comment by derrickmehaffy** [comment]
Yeah I do think we need to heavily test this against many community plugins before we consider merging. This could cause a metric ton of problems if we miss something.

Might also be good to run an extended beta here too for a few months and let people who are currently building plugins test against it also.

Also doesn't react 18 default to SSR instead of CSR? Have we tested building and ejecting the admin away from the koa instance?

Likewise if it's SSR we may have an increase in resource usage on the server and may impact our min/recommend RAM needs too.

**Comment by joshuaellis** [comment]
Vercel deployment is broken on `main` and will be fixed by #16229

File: vercel.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
     "silent": true,
     "autoJobCancelation": true
   },
-  "installCommand": "yarn setup",
+  "installCommand": "yarn install --immutable",
```
Comment (christiancp100) [inline vercel.json:7]: Why does this need to be changed?

Comment (joshuaellis) [reply vercel.json:7]: I need to rebase the target branch, this comes from `main` because on `main` we don't have setup no frozen-lockfile.

**Pull request merged**